lib/compilers/sail.ts

```diff
@@ -0,0 +1,177 @@
+        }
+
+        fullResult.executableFilename = inputFilename + '.exe';
+
+        // Query the sail compiler via `sail -dir` to find out where the
+        // C runtime files are (`sail.c` etc).
+        const sailDirResult = await this.doBuildstepAndAddToResult(
+            fullResult,
+            'Get Sail dir',
+            compiler,
+            ['-dir'],
+            execOptions,
+        );
+
+        if (sailDirResult.code !== 0) {
+            return fullResult;
+        }
+
+        const sailDir = sailDirResult.stdout.map(line => line.text).join('\n').trim();
+
+        // Now compile the C file to an executable.
+        const outputFilenameC = this.getOutputFilename(tmpDir, this.outputFilebase);
+        const outputFilenameExe = outputFilenameC + '.exe';
+
+        const compileResult = await this.doBuildstepAndAddToResult(
+            fullResult,
+            'C to binary',
+            'cc',
+            [
+                outputFilenameC,
+                // Sail C support files.
+                '-I', `${sailDir}/lib`,
+                `${sailDir}/lib/elf.c`,
+                `${sailDir}/lib/rts.c`,
+                `${sailDir}/lib/sail.c`,
+                `${sailDir}/lib/sail_failure.c`,
+                // Support for .elf.gz files. This has been removed in Sail 0.19
+                // so it isn't needed for future versions.
+                '-lz',
+                // For arbitrary precision integer types.
+                '-lgmp',
+                // Enable optimisations so the assembly isn't hilariously verbose.
+                // Ideally this would be user configurable but we'd need
+                // something like `-Wl,..` to pass options through Sail to
+                // here and that doesn't really exist.
+                '-O',
+                '-o', outputFilenameExe,
+            ],
+            execOptions,
+        );
+
+        // This is weird, but as far as I can tell CE expects the same
+        // output file name when compiling to IR (C in this case)
+        // and when compiling to a binary. If you don't do this it
+        // tries to do nonsensical things like objdumping the C, so we
+        // copy the binary back over the C file.
+        if (compileResult.code === 0 && await fs.pathExists(outputFilenameExe)) {
+            console.log(`Copying ${outputFilenameExe} to ${outputFilenameC}`);
+            await fs.copyFile(outputFilenameExe, outputFilenameC);
+        }
+
+        return fullResult;
+    }
+
+    override getOutputFilename(dirPath: string, outputFilebase: string, key?: any): string {
+        return path.join(dirPath, `${outputFilebase}.c`);
+    }
+
+    override getSharedLibraryPathsAsArguments(
+        libraries: SelectedLibraryVersion[],
+        libDownloadPath: string | undefined,
+        toolchainPath: string | undefined,
+        dirPath: string,
+    ): string[] {
+        // By default this adds -L./lib which isn't understood by the Sail compiler.
```

Comment:
We hit this with the `ispc` change I linked in the comments too :)

Reply:
Ah great, I'll check it again.